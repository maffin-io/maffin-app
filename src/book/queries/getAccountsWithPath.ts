import { FindOptionsRelations, Not } from 'typeorm';
import { Account, Book } from '../entities';

type AccountPath = {
  guid: string,
  type: string,
  path: string,
  commodity: string,
  commodityGuid: string,
};

type GetAccountsWithPathProps = {
  showRoot?: boolean,
  relations?: FindOptionsRelations<Account>,
};

export async function getAccountsWithPath(
  {
    showRoot = false,
    relations = {},
  }: GetAccountsWithPathProps = {},
): Promise<Account[]> {
  const start = performance.now();
  const books = await Book.find();
  const rootAccount = books[0].root;

  const typeFilter = !showRoot ? { type: Not('ROOT') } : {};

  const [accountsPaths, accounts] = await Promise.all([
    Account.query(
      `
      WITH RECURSIVE absolute_names(guid, name, parent_guid, path) AS (
        SELECT guid, name, parent_guid, name FROM accounts WHERE parent_guid = $1
        UNION ALL
        SELECT a.guid, a.name, a.parent_guid, p.path || ':' || a.name
        FROM accounts AS a JOIN absolute_names AS p ON a.parent_guid = p.guid
      )
      SELECT guid, path 
      FROM absolute_names
      `,
      [rootAccount.guid],
    ),
    Account.find({
      where: typeFilter,
      relations,
    }),
  ]);

  accounts.forEach((account) => {
    const accountPath = accountsPaths.find(
      (ap: AccountPath) => ap.guid === account.guid,
    );

    if (showRoot && account.type === 'ROOT') {
      account.path = 'Root';
    } else {
      account.path = accountPath.path;
    }
  });

  const end = performance.now();
  console.log(`get accounts with paths: ${end - start}ms`);

  return accounts;
}
