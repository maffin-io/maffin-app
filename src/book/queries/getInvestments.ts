import { Account } from '../entities';
import { InvestmentAccount } from '../models';
import { PriceDB, PriceDBMap } from '../prices';

export async function getInvestments(mainCurrency: string): Promise<InvestmentAccount[]> {
  const start = performance.now();
  const [accounts, todayPrices, mainCurrencyPrices] = await Promise.all([
    Account.find({
      where: [
        { type: 'STOCK' },
        { type: 'MUTUAL' },
      ],
      relations: {
        splits: {
          fk_transaction: {
            splits: {
              fk_account: true,
            },
          },
        },
      },
      order: {
        splits: {
          fk_transaction: {
            date: 'ASC',
          },
        },
      },
    }),
    PriceDB.getTodayQuotes(),
    PriceDB.getHistory(mainCurrency),
  ]);

  const pricesMap = new PriceDBMap([
    ...Object.values(mainCurrencyPrices.map),
    ...Object.values(todayPrices.map),
  ]);
  const investments = accounts.map(
    account => {
      try {
        const investment = new InvestmentAccount(
          account,
          mainCurrency,
          pricesMap,
        );
        return investment;
      } catch (e) {
        throw new Error(`Failed to create investment from account ${account.guid}: ${e}`);
      }
    },
  );
  const end = performance.now();
  console.log(`get investments: ${end - start}ms`);

  return investments;
}
