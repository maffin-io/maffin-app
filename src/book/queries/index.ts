export { getInvestments } from './getInvestments';
export { getMainCurrency } from './getMainCurrency';
export { getEarliestDate } from './getEarliestDate';
export { getSplits } from './getSplits';
export { getLatestTxs } from './getLatestTxs';
