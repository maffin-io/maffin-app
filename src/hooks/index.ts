export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
