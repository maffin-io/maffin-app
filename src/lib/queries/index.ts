export { default as getPrices } from './getPrices';

export { default as getMonthlyTotals } from './getMonthlyTotals';
export { default as getMainCurrency } from './getMainCurrency';
export { default as getEarliestDate } from './getEarliestDate';
export { default as getLatestTxs } from './getLatestTxs';
export { default as getAccountsTotals } from './getAccountsTotals';
