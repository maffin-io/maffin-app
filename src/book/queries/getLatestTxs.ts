import { Transaction } from '@/book/entities';

export async function getLatestTxs(): Promise<Transaction[]> {
  const txs = await Transaction.find({
    relations: {
      splits: {
        fk_account: true,
      },
    },
    order: {
      date: 'DESC',
    },
    take: 5,
  });

  return txs;
}
