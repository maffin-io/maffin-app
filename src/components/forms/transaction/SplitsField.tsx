import React from 'react';
import { useFieldArray } from 'react-hook-form';
import type { UseFormReturn } from 'react-hook-form';

import { Tooltip } from '@/components/tooltips';
import { Split } from '@/book/entities';
import type { FormValues } from './types';
import SplitField from './SplitField';
import MainSplit from './MainSplit';

export type SplitsFieldProps = {
  action?: 'add' | 'update' | 'delete',
  form: UseFormReturn<FormValues>,
  disabled?: boolean,
};

export default function SplitsField({
  action = 'add',
  form,
  disabled = false,
}: SplitsFieldProps): JSX.Element {
  const { fields, append, remove } = useFieldArray({
    control: form.control,
    name: 'splits',
  });

  const splits = form.watch('splits');
  const mainSplit = splits[0];
  let firstSplitIndex = 1;
  if (action === 'update') {
    firstSplitIndex = 0;
  }

  React.useEffect(() => {
    if (
      splits.length === 2
      && action === 'add'
    ) {
      if (
        splits[0].account?.commodity?.guid === splits[1].account?.commodity?.guid
      ) {
        form.setValue('splits.1.quantity', -mainSplit.quantity);
      }

      form.setValue('splits.1.value', -mainSplit.value, { shouldValidate: true });
      form.trigger('splits');
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [splits.length, mainSplit.value, form.formState.isDirty]);

  return (
    <>
      {
        action !== 'update'
        && (
          <div>
            <label className="inline-block mb-2">Amount</label>
            <MainSplit form={form} disabled={disabled} />
          </div>
        )
      }
      <label className="inline-block mt-5 mb-2">Records</label>
      {
        fields.slice(firstSplitIndex).map((item, index) => {
          index += firstSplitIndex;
          return (
            <fieldset className="grid grid-cols-12" key={item.id}>
              <div className="col-span-11">
                <SplitField
                  action={action}
                  index={index}
                  form={form}
                  disabled={disabled}
                />
              </div>
              {(
                index > 1
                && (
                  <div className="col-span-1">
                    <button
                      type="button"
                      className="btn btn-primary rounded-sm"
                      onClick={() => remove(index)}
                    >
                      X
                    </button>
                  </div>
                )
              )}
            </fieldset>
          );
        })
      }
      <p className="invalid-feedback">{form.formState.errors.splits?.message}</p>
      {
        !disabled
        && (
          <button
            className="link m-3"
            type="button"
            onClick={() => (
              append(Split.create({
                value: 0,
                quantity: 0,
              }))
            )}
          >
            Add split
            <span
              className="badge default ml-0.5"
              data-tooltip-id="add-split-help"
            >
              ?
            </span>
            <Tooltip
              id="add-split-help"
            >
              <p className="mb-2">
                Adding extra splits to a transaction is helpful when you want to split
                a transaction between different categories.
              </p>
              <p>
                An example would be a transaction where you deduct 100 euros from your
                Asset account and send 70 to Groceries and 30 to Restaurant.
              </p>
            </Tooltip>
          </button>
        )
      }
    </>
  );
}
