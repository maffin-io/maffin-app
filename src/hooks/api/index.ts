import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
