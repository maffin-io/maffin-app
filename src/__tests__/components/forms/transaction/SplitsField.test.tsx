import React from 'react';
import { useForm } from 'react-hook-form';
import { DataSource } from 'typeorm';
import userEvent from '@testing-library/user-event';
import {
  waitFor,
  render,
  screen,
} from '@testing-library/react';
import type { UseQueryResult } from '@tanstack/react-query';

import {
  Account,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import SplitsField from '@/components/forms/transaction/SplitsField';
import * as queries from '@/lib/queries';
import * as apiHook from '@/hooks/api';
import type { FormValues } from '@/components/forms/transaction/types';

jest.mock('@/lib/queries', () => ({
  __esModule: true,
  ...jest.requireActual('@/lib/queries'),
}));

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

describe('SplitsField', () => {
  let datasource: DataSource;
  beforeEach(async () => {
    jest.spyOn(apiHook, 'useAccounts').mockReturnValue(
      {
        data: undefined,
      } as UseQueryResult<Account[]>,
    );
    datasource = new DataSource({
      type: 'sqljs',
      dropSchema: true,
      entities: [Account, Commodity, Split, Transaction, Price],
      synchronize: true,
      logging: false,
    });
    await datasource.initialize();
  });

  afterEach(async () => {
    await datasource.destroy();
  });

  it('renders with default values', async () => {
    const { container } = render(<FormWrapper />);

    await screen.findByRole('combobox', { name: 'splits.1.account' });
    expect(screen.getAllByRole('spinbutton', { hidden: true })).toHaveLength(4);
    screen.getByText('Add split');

    expect(container).toMatchSnapshot();
  });

  it('renders as expected when disabled', async () => {
    render(<FormWrapper disabled />);

    expect(await screen.findByLabelText('splits.1.account')).toBeDisabled();
    expect(screen.queryByText('Add split')).toBeNull();
  });

  it('autocompletes value and quantity for split2 when only 2 splits', async () => {
    const user = userEvent.setup();
    const eur = {
      guid: 'eur',
      mnemonic: 'EUR',
      namespace: 'CURRENCY',
    } as Commodity;
    jest.spyOn(queries, 'getMainCurrency').mockResolvedValue(eur);

    jest.spyOn(apiHook, 'useAccounts').mockReturnValue(
      {
        data: [
          {
            guid: 'account_guid_1',
            path: 'path1',
            type: 'ASSET',
            commodity: eur,
          } as Account,
        ],
      } as UseQueryResult<Account[]>,
    );

    render(<FormWrapper />);

    const q0 = screen.getByRole('spinbutton', { name: 'splits.0.quantity' });
    expect(q0).toBeEnabled();

    await user.type(q0, '100');
    const v0 = screen.getByRole('spinbutton', { name: '', hidden: true });
    await waitFor(() => expect(v0).toHaveValue(100));

    await user.click(screen.getByLabelText('splits.1.account'));
    await user.click(screen.getByText('path1'));

    const q1 = screen.getByRole('spinbutton', { name: 'splits.1.quantity' });
    const v1 = screen.getByRole('spinbutton', { name: 'splits.1.value', hidden: true });
    await waitFor(() => expect(v1).toHaveValue(-100));
    await waitFor(() => expect(q1).toHaveValue(-100));
  });

  it('can remove for N > 2 split fields', async () => {
    const user = userEvent.setup();
    render(<FormWrapper />);

    await user.click(screen.getByText('Add split'));

    screen.getByRole('combobox', { name: 'splits.1.account' });
    screen.getByRole('combobox', { name: 'splits.2.account' });
    expect(screen.getAllByRole('spinbutton', { hidden: true })).toHaveLength(6);

    const removeButton = screen.getByText('X', { selector: 'button' });
    await user.click(removeButton);

    expect(screen.getAllByRole('combobox')).toHaveLength(1);
    expect(screen.getAllByRole('spinbutton', { hidden: true })).toHaveLength(4);
  });
});

function FormWrapper({ disabled = false }: { disabled?: boolean }): JSX.Element {
  const account = {
    guid: 'account_guid_1',
    path: 'path1',
    type: 'ASSET',
    commodity: {
      guid: 'eur',
      mnemonic: 'EUR',
      namespace: 'CURRENCY',
    } as Commodity,
  } as Account;

  const form = useForm<FormValues>({
    defaultValues: {
      splits: [
        {
          value: 0,
          quantity: 0,
          fk_account: account,
          account,
        },
        {},
      ],
      fk_currency: {
        guid: 'eur',
        mnemonic: 'EUR',
        namespace: 'CURRENCY',
      },
    },
  });

  return (
    <form onSubmit={form.handleSubmit(() => {})}>
      <input
        id="dateInput"
        data-testid="date"
        className="block w-full m-0"
        {...form.register('date')}
        type="date"
      />
      <SplitsField form={form} disabled={disabled} />
    </form>
  );
}
