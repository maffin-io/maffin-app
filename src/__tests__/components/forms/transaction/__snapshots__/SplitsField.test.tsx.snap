// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitsField renders with default values 1`] = `
<div>
  <form>
    <input
      class="block w-full m-0"
      data-testid="date"
      id="dateInput"
      name="date"
      type="date"
    />
    <label
      class="inline-block mb-2"
    >
      Amount
    </label>
    <div
      class="flex w-1/2"
    >
      <div
        class="flex items-center rounded-md bg-light-100 dark:bg-dark-800"
      >
        <input
          aria-label="splits.0.quantity"
          class="w-full text-right m-0"
          name="splits.0.quantity"
          step="0.001"
          type="number"
        />
        <span
          class="pr-2"
        >
          €
        </span>
      </div>
      <div
        class="flex items-center rounded-md hidden"
      >
        <span
          class="badge mx-3"
          data-tooltip-id="value-help"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m11.293 17.293 1.414 1.414L19.414 12l-6.707-6.707-1.414 1.414L15.586 11H6v2h9.586z"
            />
          </svg>
        </span>
        <div
          class="flex items-center rounded-md  bg-light-100 dark:bg-dark-800"
        >
          <input
            aria-label="splits.0.value"
            class="w-full text-right m-0"
            hidden=""
            name="splits.0.value"
            step="0.001"
            type="number"
          />
          <span
            class="pr-2"
          >
            €
          </span>
        </div>
      </div>
    </div>
    <label
      class="inline-block mt-5 mb-2"
    >
      Records
    </label>
    <fieldset
      class="grid grid-cols-12"
    >
      <div
        class="col-span-11"
      >
        <fieldset
          class="grid grid-cols-13"
        >
          <div
            class="col-span-9"
          >
            <div
              class="selector css-b62m3t-container"
              id="splits.1.account"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                class="selector__control css-13cymwt-control"
              >
                <svg
                  class="absolute text-md left-[10px]"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
                  />
                </svg>
                <div
                  class="selector__value-container css-1fdsijx-ValueContainer"
                >
                  <div
                    class="selector__placeholder css-1jqq78o-placeholder"
                    id="react-select-2-placeholder"
                  >
                    &lt;account&gt;
                  </div>
                  <div
                    class="!text-inherit selector__input-container css-qbdosj-Input"
                    data-value=""
                  >
                    <input
                      aria-activedescendant=""
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-label="splits.1.account"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class="selector__input"
                      id="react-select-2-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="selector__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="hidden selector__indicator-separator css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="selector__indicator selector__dropdown-indicator css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="splits.1.account"
                type="hidden"
                value=""
              />
            </div>
            <p
              class="invalid-feedback"
            />
          </div>
          <div
            class="col-span-4"
          >
            <div
              class="flex items-center rounded-none bg-light-100 dark:bg-dark-800 rounded-r-md"
            >
              <input
                aria-label="splits.1.quantity"
                class="w-full text-right m-0"
                name="splits.1.quantity"
                step="0.001"
                type="number"
              />
              <span
                class="pr-2"
              />
            </div>
          </div>
          <div
            class="col-span-3"
            hidden=""
          >
            <div
              class="flex items-center rounded-r-md bg-light-100 dark:bg-dark-800"
            >
              <input
                aria-label="splits.1.value"
                class="w-full text-right m-0"
                name="splits.1.value"
                step="0.001"
                type="number"
              />
              <span
                class="pr-2"
              >
                €
              </span>
            </div>
          </div>
        </fieldset>
      </div>
    </fieldset>
    <p
      class="invalid-feedback"
    />
    <button
      class="link m-3"
      type="button"
    >
      Add split
    </button>
  </form>
</div>
`;
